File: Labs/Lab03-Camera-Calibration/02-calibrate_checkerboard/cpp/main.cpp
#include <iostream>
#include <opencv2/opencv.hpp>
#include <opencv2/calib3d/calib3d.hpp>
#include <opencv2/calib3d/calib3d_c.h>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>


using namespace cv;
using namespace std;

// Define the dimensions of checkerboard
int CHECKERBOARD[2] = {8, 11}; // width, height

int main() {
    // Create vector to store vectors of 3D points for each checkerboard image
    vector<vector<Point3f>> objPoints;

    // Create vector to store vectors of 2D points for each checkerboard image
    vector<vector<Point2f>> imgPoints;

    // Define the world coordinates for 3D points
    vector<Point3f> objPoint;
    for (int i=0; i < CHECKERBOARD[1]; i++)
    {
        for (int j = 0; j < CHECKERBOARD[0]; j++)
        {
            objPoint.emplace_back(j, i, 0);

        }
    }

    // Extract path of individual image stored in a given directory
    vector<String> images;

    // Path of the folder containing checkerboard images
    string path = "../../../../Data/Lab03/images/*.jpg";

    glob(path, images);

    Mat frame, gray;

    // Vector to store the pixel coordinates of detected checkerboard corners
    vector<Point2f> corner_pts;
    bool success;

    // Looping over all the images in the directory
    for (auto & image : images)
    {
        frame = imread(image);
        cvtColor(frame, gray, COLOR_BGR2GRAY);

        // Finding checkerboard corners
        // If desired number of corners are found in the image, success will be true
        success = findChessboardCorners(gray, Size(CHECKERBOARD[0], CHECKERBOARD[1]), corner_pts, CV_CALIB_CB_ADAPTIVE_THRESH | CV_CALIB_CB_FAST_CHECK | CV_CALIB_CB_NORMALIZE_IMAGE);

        /*
         * If desired number of corner are detected, refine the pixel coordinates and
         * display them on the images of checkerboard
         */
        if (success)
        {
            TermCriteria criteria(CV_TERMCRIT_EPS | CV_TERMCRIT_ITER, 30, 0.001);

            // Refine pixel coordinates for given 2D points
            cornerSubPix(gray, corner_pts, Size(11, 11), Size(-1, -1), criteria);

            // Display the detected corner points on the checkerboard
            drawChessboardCorners(frame, Size(CHECKERBOARD[0], CHECKERBOARD[1]), corner_pts, true);

            objPoints.push_back(objPoint);
            imgPoints.push_back(corner_pts);
        }

        imshow("Image", frame);
        waitKey(0);
    }
    destroyAllWindows();

    Mat cameraMatrix, distCoefficients, R, T;

    /*
     * Perform camera calibration by passing the value of known 3D points (objPoints)
     * and corresponding pixel coordinates of the detected corners (imgPoints)
     */
    calibrateCamera(objPoints, imgPoints, Size(gray.rows, gray.cols),
                    cameraMatrix, distCoefficients, R, T);
    cout << "Camera Matrix: " << cameraMatrix << endl << endl;
    cout << "Distortion Coefficient: " << distCoefficients << endl << endl;
    cout << "Rotation Vector: " << R << endl << endl;
    cout << "Translation Vector: " << T << endl << endl;

    return 0;
}

File: Labs/Lab03-Camera-Calibration/02-calibrate_checkerboard/python/main.py
import glob
import cv2 as cv
import numpy as np


# Define the dimensions of checkerboard
CHECKERBOARD = (8, 11)
criteria = (cv.TERM_CRITERIA_EPS + cv.TERM_CRITERIA_MAX_ITER, 30, 0.001)

# Create vector to store vectors of 3D points for each checkerboard image
obj_points = []
# Create vector to store vectors of 2D points for each checkerboard image
img_points = []

# Define the world coordinates for 3D points
obj_point = np.zeros((1, CHECKERBOARD[0] * CHECKERBOARD[1], 3), dtype=np.float32)
obj_point[0, :, :2] = np.mgrid[0:CHECKERBOARD[0], 0:CHECKERBOARD[1]].T.reshape(-1, 2)
prev_img_shape = None

# Extracting path of individual image stored in a given directory
images = glob.glob("../../../Data/Lab03/checkerboard/images/*.jpg")
img = None
gray = None
for f_name in images:

    img = cv.imread(f_name)
    gray = cv.cvtColor(img, cv.COLOR_BGR2GRAY)

    # Find the chessboard corners
    # If desired number of corners are found in the image then ret = true
    ret, corners = cv.findChessboardCorners(gray, CHECKERBOARD, cv.CALIB_CB_ADAPTIVE_THRESH + cv.CALIB_CB_FAST_CHECK
                                            + cv.CALIB_CB_NORMALIZE_IMAGE)
    """
        If desired number of corner are detected, refine the pixel coordinates and display them 
        on the images of checkerboard
    """
    if ret:
        obj_points.append(obj_point)

        # Refining pixel coordinates for given 2D points.
        corners2 = cv.cornerSubPix(gray, corners, (11, 11), (-1, -1), criteria)

        img_points.append(corners2)

        # Draw and display the corners
        img = cv.drawChessboardCorners(img, CHECKERBOARD, corners2, ret)
    cv.imshow('img', img)
    cv.waitKey(0)

cv.destroyAllWindows()

h, w = img.shape[:2]
"""
    Performing camera calibration by passing the value of known 3D points (obj_points) and 
    corresponding pixel coordinates of the detected corners (img_points)
"""
ret, mtx, dist, r_vectors, t_vectors = cv.calibrateCamera(obj_points, img_points, gray.shape[::-1], None, None)

print("Camera matrix: \n")
print(mtx)
print("Distortion Coefficient: \n")
print(dist)
print("Rotation vectors: \n")
print(r_vectors)
print("Translation vectors: \n")
print(t_vectors)

# Show images of undistorted
for f_name in images:
    img = cv.imread(f_name)
    res = cv.undistort(img, mtx, dist)
    cv.imshow('img', img)
    cv.imshow('res', res)
    cv.waitKey(0)

# Use ROI obtained above to crop the result
print("Undistorted using ROI")
for f_name in images:
    print(f_name)

    img = cv.imread(f_name)

    h, w = img.shape[:2]
    new_camera_mtx, roi = cv.getOptimalNewCameraMatrix(mtx, dist, (w, h), 1, (w, h))
    res = cv.undistort(img, mtx, dist)

    # crop the image
    x, y, w, h = roi
    res = res[y:y+h, x:x+w]

    cv.imshow('img', img)
    cv.imshow('res', res)
    cv.waitKey(0)


# Find a mapping function from the distorted image to undistorted image.
# Then use the remap function.
print("Find mapping from distorted to undistorted image.\nThen use the remap function.")
for f_name in images:
    print(f_name)

    img = cv.imread(f_name)
    h, w = img.shape[:2]
    new_camera_mtx, roi = cv.getOptimalNewCameraMatrix(mtx, dist, (w, h), 1, (w, h))

    map_x, map_y = cv.initUndistortRectifyMap(mtx, dist, None, new_camera_mtx, (w, h), 5)
    res = cv.remap(img, map_x, map_y, cv.INTER_LINEAR)

    # Crop the image
    x, y, w, h = roi
    res = res[y:y+h, x:x+w]

    cv.imshow('img', img)
    cv.imshow('res', res)
    cv.waitKey(0)

mean_error = 0
for i in range(len(obj_points)):
    img_points2, _ = cv.projectPoints(obj_points[i], r_vectors[i], t_vectors[i], mtx, dist)
    error = cv.norm(img_points[i], img_points2, cv.NORM_L2) / len(img_points2)
    mean_error += error
print(f"Total error: {mean_error / len(obj_points)}")
